platform/posix/transport/src/openssl_posix.c

```diff
@@ -726,9 +726,13 @@ int32_t Openssl_Recv( NetworkContext_t * pNetworkContext,
         {
             sslError = SSL_get_error( pOpensslParams->pSsl, bytesReceived );
 
-            if( sslError == SSL_ERROR_WANT_READ )
+            if( ( sslError == SSL_ERROR_WANT_READ ) || ( sslError == SSL_ERROR_WANT_WRITE ) )
             {
-                /* There is no data to receive at this time. */
+                /* The OpenSSL documentation mentions that SSL_Write can have a return code of
+                 * SSL_ERROR_WANT_READ and SSL_ERROR_WANT_WRITE. Both of those codes mean that the
+                 * SSL_read() operation needs to be retried to complete the read operation.
+                 * Thus, setting the return value of this function as zero to represent that no
+                 * data was received from the network. */
```

Comment:
@abhidixi11 and I had a discussion a while back about this block, and I believe that the `else` block below this should set `bytesReceived` to a negative value, otherwise it's possible that a non-retriable error may return 0.

Reply:
Although, I don't think we should necessarily change the error code returned by OpenSSL. We could probably have an else if like so:
```
else if( bytesReceived = 0 || sslError == SSL_ERROR_ZERO_RETURN )
```

Reply:
I thought about that and wasn't sure about what negative value this should be set to. The expectation by the calling code could be to expect the negative error code sent by OpenSSL.
However, for the correctness in functional behavior of the transport send/recv functions, I will set the value to be explicitly negative as `-1`.

Reply:
I think the negative error code sent by OpenSSL should be returned. The only exception to this case is when OpenSSL returns `0`, which signifies a close notify, then transport recv should return `-1`.

Reply:
[According to the documentation](https://www.openssl.org/docs/man1.1.1/man3/SSL_read.html), this is not as simple as returning `SSL_ERROR_WANT_READ` in blocking mode.  In order to get that return value, one of the following must be true:

* Underlying BIO is non-blocking (this doesn't seem to be the case based on later comments).  In that case you also need to check for `SSL_ERROR_WANT_READ`.
* Underlying BIO is blocking __and__ `SSL_MODE_AUTO_RETRY` is disabled.

These comments and the initialization/configuration of this TLS need to be very clear and consistent with the OpenSSL documentation, so let's clarify this.